pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table.go

```diff
@@ -119,33 +125,53 @@ func alterTableIsSupported(n *tree.AlterTable, mode sessiondatapb.NewSchemaChang
 	for _, cmd := range n.Cmds {
-		minSupportedClusterVersion := supportedAlterTableStatements[reflect.TypeOf(cmd)].minSupportedClusterVersion
-		if !b.EvalCtx().Settings.Version.IsActive(b, minSupportedClusterVersion) {
-			return false
+		if addConstraint, isAddConstraint := cmd.(*tree.AlterTableAddConstraint); isAddConstraint {
+			if !alterTableAddConstraintSupportedInCurrentClusterVersion(b, addConstraint) {
+				return false
+			}
+		} else {
+			minSupportedClusterVersion := supportedAlterTableStatements[reflect.TypeOf(cmd)].minSupportedClusterVersion
+			if !b.EvalCtx().Settings.Version.IsActive(b, minSupportedClusterVersion) {
+				return false
+			}
 		}
 	}
 	return true
 }
 
-// AlterTable implements ALTER TABLE.
-func AlterTable(b BuildCtx, n *tree.AlterTable) {
-	// Hoist the constraints to separate clauses because other code assumes that
-	// that is how the commands will look.
-	n.HoistAddColumnConstraints(func() {
-		telemetry.Inc(sqltelemetry.SchemaChangeAlterCounterWithExtra("table", "add_column.references"))
-	})
-
-	// We already confirmed the command is supported, but run the extra checks for
-	// it now.
-	for _, cmd := range n.Cmds {
-		// Run the extraChecks to see if we should avoid even resolving the
-		// descriptor.
-		info := supportedAlterTableStatements[reflect.TypeOf(cmd)]
-		if info.on && info.extraChecks != nil && !reflect.ValueOf(info.extraChecks).
-			Call([]reflect.Value{reflect.ValueOf(cmd)})[0].Bool() {
-			panic(scerrors.NotImplementedError(cmd))
+// alterTableAddConstraintSupportedInCurrentClusterVersion determines if a particular
+// AlterTableAddConstraint command is supported under the current cluster version.
+func alterTableAddConstraintSupportedInCurrentClusterVersion(
+	b BuildCtx, constraint *tree.AlterTableAddConstraint,
+) bool {
+	var cmdKey string
+	// Figure out alter table add constraint command type: PRIMARY_KEY, CHECK, FOREIGN_KEY, or UNIQUE
+	switch d := constraint.ConstraintDef.(type) {
+	case *tree.UniqueConstraintTableDef:
+		if d.PrimaryKey {
+			cmdKey = "ADD_PRIMARY_KEY"
+		} else {
+			panic(errors.AssertionFailedf("previously checked guardrail violated: Unsupported add constraint command type"))
 		}
+	default:
+		panic(errors.AssertionFailedf("previously checked guardrail violated: Unknown add constraint command type"))
 	}
+	// Figure out command validation behavior: DEFAULT or SKIP
+	switch constraint.ValidationBehavior {
+	case tree.ValidationDefault:
+		cmdKey += "_DEFAULT"
+	case tree.ValidationSkip:
+		panic(errors.AssertionFailedf("previously checked guardrail violated: unsupported validation behavior"))
+	default:
+		panic(errors.AssertionFailedf("previously checked guardrail violated: unknown validation behavior"))
+	}
+
+	minSupportedClusterVersion := alterTableAddConstraintMinSupportedClusterVersion[cmdKey]
+	return b.EvalCtx().Settings.Version.IsActive(b, minSupportedClusterVersion)
+}
```

Comment:
I'm sorry but this code is still needlessly complex. Rather than dispatching in a switch block why not introduce a `supportedAlterTableAddConstraintStatements` map in the same vein as the other maps? That way you can get rid of all these useless `panic(errors.AssertionFailedf("previously checked guardrail violated: Unknown add constraint command type"))` which are effectively dead code. Also instead of keying on free-form strings, key on the `reflect.Type` of an AST node.